File: src/test/java/com/github/nsc/de/shake/parser/LogicalTests.java
package com.github.nsc.de.shake.parser;

import com.github.nsc.de.shake.parser.node.expression.AddNode;
import com.github.nsc.de.shake.parser.node.factor.DoubleNode;
import com.github.nsc.de.shake.parser.node.factor.IntegerNode;
import com.github.nsc.de.shake.parser.node.expression.MulNode;
import com.github.nsc.de.shake.parser.node.logical.*;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;
import static com.github.nsc.de.shake.TestUtil.*;
import static com.github.nsc.de.shake.parser.ParserTestUtil.*;

public class LogicalTests {

    @Test
    public void testTrue() {
        assertNotNull(parseSingle("<LogicalTrueTest>","true", LogicalTrueNode.class));
    }

    @Test
    public void testFalse() {
        assertNotNull(parseSingle("<LogicalFalseTest>","false", LogicalFalseNode.class));
    }

    @Test
    public void testEqEquals() {
        testBasic("10==10", 10, 10, LogicalEqEqualsNode.class);
    }

    @Test
    public void testBiggerEquals() {
        testBasic("10>=10", 10, 10, LogicalBiggerEqualsNode.class);
    }

    @Test
    public void testSmallerEquals() {
        testBasic("10<=10", 10, 10, LogicalSmallerEqualsNode.class);
    }

    @Test
    public void testBigger() {
        testBasic("10>10", 10, 10, LogicalBiggerNode.class);
    }

    @Test
    public void testSmaller() {
        testBasic("10<10", 10, 10, LogicalSmallerNode.class);
    }

    @Test
    public void testAnd() {

        LogicalAndNode node = parseSingle("<LogicalAndTest>", "true && false", LogicalAndNode.class);
        assertNotNull(node.getLeft());
        assertType(LogicalTrueNode.class, node.getLeft());
        assertNotNull(node.getRight());
        assertType(LogicalFalseNode.class, node.getRight());

    }

    @Test
    public void testOr() {

        LogicalOrNode node = parseSingle("<LogicalOrTest>", "true || false", LogicalOrNode.class);
        assertNotNull(node.getLeft());
        assertType(LogicalTrueNode.class, node.getLeft());
        assertNotNull(node.getRight());
        assertType(LogicalFalseNode.class, node.getRight());

    }

    @Test
    public void testBrackets() {

        LogicalAndNode node = parseSingle("<LogicalBracketTest>", "true && (false || true)", LogicalAndNode.class);
        assertNotNull(node.getLeft());
        assertType(LogicalTrueNode.class, node.getLeft());

        assertNotNull(node.getRight());
        assertType(LogicalOrNode.class, node.getRight());

        LogicalOrNode or = (LogicalOrNode) node.getRight();
        assertNotNull(or.getLeft());
        assertType(LogicalFalseNode.class, or.getLeft());
        assertNotNull(or.getRight());
        assertType(LogicalTrueNode.class, or.getRight());

    }

    @Test
    public void testExpr() {

        LogicalBiggerEqualsNode node = parseSingle("<LogicalExpressionTest>", "10 >= 5 + 9 * 2", LogicalBiggerEqualsNode.class);
        assertNotNull(node.getLeft());
        assertType(IntegerNode.class, node.getLeft());
        assertEquals(10, ((IntegerNode) node.getLeft()).getNumber());
        assertNotNull(node.getRight());
        assertType(AddNode.class, node.getRight());

        AddNode add = (AddNode) node.getRight();
        assertNotNull(add.getLeft());
        assertType(IntegerNode.class, add.getLeft());
        assertEquals(5, ((IntegerNode) add.getLeft()).getNumber());
        assertNotNull(add.getRight());
        assertType(MulNode.class, add.getRight());

        MulNode mul = (MulNode) add.getRight();
        assertNotNull(mul.getLeft());
        assertType(IntegerNode.class, mul.getLeft());
        assertEquals(9, ((IntegerNode) mul.getLeft()).getNumber());
        assertNotNull(mul.getRight());
        assertType(IntegerNode.class, mul.getRight());
        assertEquals(2, ((IntegerNode) mul.getRight()).getNumber());

    }



    private <T extends LogicalCompareNode> void testBasic(String input, double left, double right, Class<T> type) {

        T node = parseSingle('<'+type.getSimpleName().substring(type.getSimpleName().length() - 4)+"Test>", input, type);
        assertNotNull(node.getLeft());
        assertTrue(node.getLeft() instanceof DoubleNode || node.getLeft() instanceof IntegerNode);
        if(node.getLeft() instanceof DoubleNode) assertEquals(left, ((DoubleNode) node.getLeft()).getNumber());
        else assertEquals(left, ((IntegerNode) node.getLeft()).getNumber());
        assertNotNull(node.getRight());
        assertTrue(node.getRight() instanceof DoubleNode || node.getRight() instanceof IntegerNode);
        if(node.getRight() instanceof DoubleNode) assertEquals(right, ((DoubleNode) node.getRight()).getNumber());
        else assertEquals(right, ((IntegerNode) node.getRight()).getNumber());

    }
}

File: parser/src/test/kotlin/com/github/nsc/de/shake/parser/LogicalTests.kt
package com.github.nsc.de.shake.parser

import com.github.nsc.de.shake.assertType
import com.github.nsc.de.shake.parser.node.expression.AddNode
import com.github.nsc.de.shake.parser.node.expression.MulNode
import com.github.nsc.de.shake.parser.node.factor.DoubleNode
import com.github.nsc.de.shake.parser.node.factor.IntegerNode
import com.github.nsc.de.shake.parser.node.logical.*
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class LogicalTests {
    @Test
    fun testTrue() {
        Assertions.assertNotNull(ParserTestUtil.parseSingle("<LogicalTrueTest>", "true", LogicalTrueNode::class.java))
    }

    @Test
    fun testFalse() {
        Assertions.assertNotNull(
            ParserTestUtil.parseSingle(
                "<LogicalFalseTest>",
                "false",
                LogicalFalseNode::class.java
            )
        )
    }

    @Test
    fun testEqEquals() {
        testBasic("10==10", 10.0, 10.0, LogicalEqEqualsNode::class.java)
    }

    @Test
    fun testBiggerEquals() {
        testBasic("10>=10", 10.0, 10.0, LogicalBiggerEqualsNode::class.java)
    }

    @Test
    fun testSmallerEquals() {
        testBasic("10<=10", 10.0, 10.0, LogicalSmallerEqualsNode::class.java)
    }

    @Test
    fun testBigger() {
        testBasic("10>10", 10.0, 10.0, LogicalBiggerNode::class.java)
    }

    @Test
    fun testSmaller() {
        testBasic("10<10", 10.0, 10.0, LogicalSmallerNode::class.java)
    }

    @Test
    fun testAnd() {
        val node = ParserTestUtil.parseSingle("<LogicalAndTest>", "true && false", LogicalAndNode::class.java)
        Assertions.assertNotNull(node.left)
        assertType(LogicalTrueNode::class.java, node.left)
        Assertions.assertNotNull(node.right)
        assertType(LogicalFalseNode::class.java, node.right)
    }

    @Test
    fun testOr() {
        val node = ParserTestUtil.parseSingle("<LogicalOrTest>", "true || false", LogicalOrNode::class.java)
        Assertions.assertNotNull(node.left)
        assertType(LogicalTrueNode::class.java, node.left)
        Assertions.assertNotNull(node.right)
        assertType(LogicalFalseNode::class.java, node.right)
    }

    @Test
    fun testXOr() {
        val node = ParserTestUtil.parseSingle("<LogicalXOrTest>", "true ^ false", LogicalXOrNode::class.java)
        Assertions.assertNotNull(node.left)
        assertType(LogicalTrueNode::class.java, node.left)
        Assertions.assertNotNull(node.right)
        assertType(LogicalFalseNode::class.java, node.right)
    }

    @Test
    fun testBrackets() {
        val node =
            ParserTestUtil.parseSingle("<LogicalBracketTest>", "true && (false || true)", LogicalAndNode::class.java)
        Assertions.assertNotNull(node.left)
        assertType(LogicalTrueNode::class.java, node.left)
        Assertions.assertNotNull(node.right)
        assertType(LogicalOrNode::class.java, node.right)
        val or = node.right as LogicalOrNode
        Assertions.assertNotNull(or.left)
        assertType(LogicalFalseNode::class.java, or.left)
        Assertions.assertNotNull(or.right)
        assertType(LogicalTrueNode::class.java, or.right)
    }

    @Test
    fun testExpr() {
        val node = ParserTestUtil.parseSingle(
            "<LogicalExpressionTest>",
            "10 >= 5 + 9 * 2",
            LogicalBiggerEqualsNode::class.java
        )
        Assertions.assertNotNull(node.left)
        assertType(IntegerNode::class.java, node.left)
        Assertions.assertEquals(10, (node.left as IntegerNode).number)
        Assertions.assertNotNull(node.right)
        assertType(AddNode::class.java, node.right)
        val add = node.right as AddNode
        Assertions.assertNotNull(add.left)
        assertType(IntegerNode::class.java, add.left)
        Assertions.assertEquals(5, (add.left as IntegerNode).number)
        Assertions.assertNotNull(add.right)
        assertType(MulNode::class.java, add.right)
        val mul = add.right as MulNode
        Assertions.assertNotNull(mul.left)
        assertType(IntegerNode::class.java, mul.left)
        Assertions.assertEquals(9, (mul.left as IntegerNode).number)
        Assertions.assertNotNull(mul.right)
        assertType(IntegerNode::class.java, mul.right)
        Assertions.assertEquals(2, (mul.right as IntegerNode).number)
    }

    private fun <T : LogicalCompareNode?> testBasic(input: String, left: Double, right: Double, type: Class<T>) {
        val node = ParserTestUtil.parseSingle(
            '<'.toString() + type.simpleName.substring(type.simpleName.length - 4) + "Test>",
            input,
            type
        )
        Assertions.assertNotNull(node!!.left)
        Assertions.assertTrue(node.left is DoubleNode || node.left is IntegerNode)
        if (node.left is DoubleNode) Assertions.assertEquals(
            left,
            (node.left as DoubleNode).number
        ) else Assertions.assertEquals(left, (node.left as IntegerNode).number.toDouble())
        Assertions.assertNotNull(node.right)
        Assertions.assertTrue(node.right is DoubleNode || node.right is IntegerNode)
        if (node.right is DoubleNode) Assertions.assertEquals(
            right,
            (node.right as DoubleNode).number
        ) else Assertions.assertEquals(right, (node.right as IntegerNode).number.toDouble())
    }
}